controllers/snapshot/snapshot_adapter.go

```diff
@@ -0,0 +1,444 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package snapshot
+
+import (
+	"context"
+	"github.com/go-logr/logr"
+	hasv1alpha1 "github.com/redhat-appstudio/application-service/api/v1alpha1"
+	"github.com/redhat-appstudio/integration-service/api/v1alpha1"
+	"github.com/redhat-appstudio/integration-service/controllers/results"
+	appstudioshared "github.com/redhat-appstudio/managed-gitops/appstudio-shared/apis/appstudio.redhat.com/v1alpha1"
+	releasev1alpha1 "github.com/redhat-appstudio/release-service/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"math"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Adapter holds the objects needed to reconcile a Release.
+type Adapter struct {
+	snapshot    *appstudioshared.ApplicationSnapshot
+	application *hasv1alpha1.Application
+	component   *hasv1alpha1.Component
+	logger      logr.Logger
+	client      client.Client
+	context     context.Context
+}
+
+// NewAdapter creates and returns an Adapter instance.
+func NewAdapter(snapshot *appstudioshared.ApplicationSnapshot, application *hasv1alpha1.Application, component *hasv1alpha1.Component, logger logr.Logger, client client.Client,
+	context context.Context) *Adapter {
+	return &Adapter{
+		snapshot:    snapshot,
+		application: application,
+		component:   component,
+		logger:      logger,
+		client:      client,
+		context:     context,
+	}
+}
+
+// EnsureAllIntegrationTestPipelinesExist is an operation that will ensure that all Integration test pipelines
+// associated with the ApplicationSnapshot and the Application's IntegrationTestScenarios exist.
+// Otherwise, it will create new Releases for each ReleasePlan.
+func (a *Adapter) EnsureAllIntegrationTestPipelinesExist() (results.OperationResult, error) {
+	if meta.FindStatusCondition(a.snapshot.Status.Conditions, "HACBSTestSucceeded") != nil {
```

Comment:
This checks could be extracted to the same file where you add the `mark` functions and become a single call to a function `haveTestsFinished` or something similar.

Reply:
Ack, extracted all functions belonging to Mark to a separate helper file and added the functions that check the status conditions